### DISCUSSION TITLE
Random reboots with assert failed: multi_heap_get_allocated_size multi_heap_poisoning.c:357 (head != NULL)

### DISCUSSION THREAD

**Discussion by Metatron135** [PlummersSoftwareLLC/NightDriverStrip#241] [General]
Hello colored light lovers! If someone is having the same issue, could you please assist.

I've just uploaded the library into the ESP32 and everything works fine except reboots. Sometimes it reboots 15 times in a row before stabilizing and sometimes it just runs and then reboots in 30 secs.

Would really appreciate some input as I'm a hobby level programmer, if I can call myself that, and the memory allocations and heaps is very vague concept for me. Thank you for your time and support!

Here is the error log:

Please build project in debug configuration to get more details about an exception.
See https://docs.platformio.org/page/projectconf/build_configurations.html

--- Terminal on /dev/cu.usbserial-0001 | 115200 8-N-1
--- Available filters and text transformations: colorize, debug, default, direct, esp32_exception_decoder, hexlify, log2file, nocontrol, printable, send_on_enter, time
--- More details at https://bit.ly/pio-monitor-filters
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H
Replacing Idle Tasks with TaskManager...
(I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
(I) 
(W) (setup)(C1) Reserving 39 LED buffers for a total of 71448 bytes...
(I) (setup)(C1) Adding LEDs to FastLED...
(I) (setup)(C1) Adding 600 LEDs to FastLED.
(I) (setup)(C1) Initializing effects manager...
(W) (InitEffectsManager)(C1) InitEffectsManager...
(I) (setup)(C1) Launching Drawing:
(E) (setup)(C1) Heap before launch: PASS
(W) (StartDrawThread)(C1) >> Launching Draw Thread
(W) (DrawLRemoteThread)(C1) >> Launching Remote Thread(W) 
(W) (Setup)(C1) >> GFXBase::Setup
(W) 
(W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
m(W) (RemoteLoopEntry(SCt0) >> RemoteLoopEntry
(W) 
(W) (begin)(C0) Remote Control Decoding Started
W) (begin)(C0) Remote Control Decodi(W) getworkHandlingLoopEn(StartSocketThreadkHa)n(dCl1i)ngLoopEntry
(W) 
[   225][E][ESPmDNS.cpp:65] begin(): Failed starting MDNS
Error starting mDNS
W) 
t Thread
(I) (setup)(C1) Setup complete - ESP32 Free Memory: 156200
(I) 
(I) (ConnectToWiFi)(C0) Setting host name to NightDriverStrip...WL_NO_SHIELD
Pass 0 of 5: Connecting to Wifi SSID: ZEN Light - ESP32 Free Memory: 156176, PSRAM:0, PSRAM Free: 0

Brownout detector was triggered

ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13132
load:0x40080400,len:3036
entry 0x400805e4
Replacing Idle Tasks with TaskManager...
(I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
(I) 
(W) (setup)(C1) Reserving 39 LED buffers for a total of 71448 bytes...
(I) (setup)(C1) Adding LEDs to FastLED...
(I) (setup)(C1) Adding 600 LEDs to FastLED.
(I) (setup)(C1) Initializing effects manager...
(W) (InitEffectsManager)(C1) InitEffectsManager...
(I) (setup)(C1) Launching Drawing:
(E) (setup)(C1) Heap before launch: PASS
(W) (StartDrawThread)(C1) >> Launching Draw Thread
(W) (DrawLRemoteThread)(C1) >> Launching Remote Thread(W) 
(W) (Setup)(C1) >> GFXBase::Setup
(W) 
(W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
m(W) (RemoteLoopEntry(SCt0) >> RemoteLoopEntry
(W) 
(W) (begin)(C0) Remote Control Decoding Started
W) (begin)(C0) Remote Control DecodiCORRUPT HEAP: Bad head at 0x(W)3ffb4f08. Expected 0xabba1234 got 0x3ffb0014

assert failed: multi_heap_get_allocated_size multi_heap_poisoning.c:357 (head != NULL)

Backtrace:0x40084399:0x3ffb23900x4008e0d9:0x3ffb23b0 0x40093961:0x3ffb23d0 0x40093637:0x3ffb2500 0x400848a4:0x3ffb2520 0x400848dd:0x3ffb2540 0x4008490f:0x3ffb2560 0x40093981:0x3ffb2580 0x400f44ab:0x3ffb25a0 0x400f451b:0x3ffb25e0 0x400f47ce:0x3ffb2600 0x400f48f5:0x3ffb2620 0x400efc21:0x3ffb2650 0x4019c061:0x3ffb26c0 0x400f414b:0x3ffb26e0 0x400d7dd3:0x3ffb2780 0x400f6b02:0x3ffb2820 

  #0  0x40084399:0x3ffb2390 in panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:402
  #1  0x4008e0d9:0x3ffb23b0 in esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:128
  #2  0x40093961:0x3ffb23d0 in __assert_func at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/assert.c:85
  #3  0x40093637:0x3ffb2500 in multi_heap_get_allocated_size at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/multi_heap_poisoning.c:357 (discriminator 1)
  #4  0x400848a4:0x3ffb2520 in heap_caps_realloc_base at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/heap_caps.c:409
      (inlined by) heap_caps_realloc_base at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/heap_caps.c:347
  #5  0x400848dd:0x3ffb2540 in heap_caps_realloc at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/heap_caps.c:423
  #6  0x4008490f:0x3ffb2560 in heap_caps_realloc_default at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/heap_caps.c:225
  #7  0x40093981:0x3ffb2580 in realloc at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/heap.c:34
  #8  0x400f44ab:0x3ffb25a0 in String::changeBuffer(unsigned int) at /Users/metatron/.platformio/packages/framework-arduinoespressif32/cores/esp32/WString.cpp:218 (discriminator 4)
  #9  0x400f451b:0x3ffb25e0 in String::reserve(unsigned int) at /Users/metatron/.platformio/packages/framework-arduinoespressif32/cores/esp32/WString.cpp:182 (discriminator 4)
  #10 0x400f47ce:0x3ffb2600 in String::concat(char const*, unsigned int) at /Users/metatron/.platformio/packages/framework-arduinoespressif32/cores/esp32/WString.cpp:367
  #11 0x400f48f5:0x3ffb2620 in String::concat(char) at /Users/metatron/.platformio/packages/framework-arduinoespressif32/cores/esp32/WString.cpp:387
  #12 0x400efc21:0x3ffb2650 in RemoteDebug::write(unsigned char) at .pio/libdeps/zv/RemoteDebug/src/RemoteDebug.cpp:1174
  #13 0x4019c061:0x3ffb26c0 in RemoteDebug::write(unsigned char const*, unsigned int) at .pio/libdeps/zv/RemoteDebug/src/RemoteDebug.cpp:860 (discriminator 2)
  #14 0x400f414b:0x3ffb26e0 in Print::printf(char const*, ...) at /Users/metatron/.platformio/packages/framework-arduinoespressif32/cores/esp32/Print.cpp:70
  #15 0x400d7dd3:0x3ffb2780 in NightDriverTaskManager::StartSocketThread() at include/taskmgr.h:258
      (inlined by) setup() at src/main.cpp:807
  #16 0x400f6b02:0x3ffb2820 in loopTask(void*) at /Users/metatron/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:42

ELF file SHA256: 0000000000000000

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13132
load:0x40080400,len:3036
entry 0x400805e4
Replacing Idle Tasks with TaskManager...
(I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
(I) 
(W) (setup)(C1) Reserving 39 LED buffers for a total of 71448 bytes...
(I) (setup)(C1) Adding LEDs to FastLED...
(I) (setup)(C1) Adding 600 LEDs to FastLED.
(I) (setup)(C1) Initializing effects manager...
(W) (InitEffectsManager)(C1) InitEffectsManager...
(I) (setup)(C1) Launching Drawing:
(E) (setup)(C1) Heap before launch: PASS
(W) (StartDrawThread)(C1) >> Launching Draw Thread
(W) (DrawLRemoteThread)(C1) >> Launching Remote Thread(W) 
(W) (Setup)(C1) >> GFXBase::Setup
(W) 
(W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
m(W) (RemoteLoopEntry(SCt0) >> RemoteLoopEntry
(W) 
(W) (begin)(C0) Remote Control Decoding Started
W) (begin)(C0) Remote Control Decodi(W) getworkHandlingLoopEn(tSrtya)r(tCS0ocketThreadkHandlingLoopEntry
(W) 
[   225][E][ESPmDNS.cpp:65] begin(): Failed starting MDNS
Error starting mDNS
W) 
t Thread
(I) (setup)(C1) Setup complete - ESP32 Free Memory: 156200
(I) 
(I) (ConnectToWiFi)(C0) Setting host name to NightDriverStrip...WL_NO_SHIELD
Pass 0 of 5: Connecting to Wifi SSID: ZEN Light - ESP32 Free Memory: 156176, PSRAM:0, PSRAM Free: 0

Brownout detector was triggered

ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13132
load:0x40080400,len:3036
entry 0x400805e4
Replacing Idle Tasks with TaskManager...
(I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
(I) 
(W) (setup)(C1) Reserving 39 LED buffers for a total of 71448 bytes...
(I) (setup)(C1) Adding LEDs to FastLED...
(I) (setup)(C1) Adding 600 LEDs to FastLED.
(I) (setup)(C1) Initializing effects manager...
(W) (InitEffectsManager)(C1) InitEffectsManager...
(I) (setup)(C1) Launching Drawing:
(E) (setup)(C1) Heap before launch: PASS
(W) (StartDrawThread)(C1) >> Launching Draw Thread
(W) (DrawLRemoteThread)(C1) >> Launching Remote Thread(W) 
(W) (Setup)(C1) >> GFXBase::Setup
(W) 
(W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
m(W) (RemoteLoopEntry(SCt0) >> RemoteLoopEntry
(W) 
(W) (begin)(C0) Remote Control Decoding Started
W) (begin)(C0) Remote Control Decodi(W) getworkHandlingLoopEn(tSrtya)r(tSocketThreadrkHandlingLoopEntry
(W) 
[   225][E][ESPmDNS.cpp:65] begin(): Failed starting MDNS
Error starting mDNS
W) 
t Thread
(I) (setup)(C1) Setup complete - ESP32 Free Memory: 156200
(I) 
(I) (ConnectToWiFi)(C0) Setting host name to NightDriverStrip...WL_NO_SHIELD

**Comment by Metatron135** [comment]
Here is my platformio.ini config:

[env:zv]
extends         = dev_esp32
build_flags     = -DZV=1
                  -DENABLE_WIFI=1
                  -DENABLE_WEBSERVER=1
                  -Ofast 
                  ${dev_esp32.build_flags}

Globals.h config:

#elif ZV 

    // This is a simple demo configuration.  To build, simply connect the data lead from a WS2812B
    // strip to pin 5.  This does not use the OLED, LCD, or anything fancy, it simply drives the
    // LEDs with a simple rainbow effect as specified in effects.cpp for DEMO.
    //
    // Please ensure you supply sufficent power to your strip, as even the DEMO of 144 LEDs, if set
    // to white, would overload a USB port.
    #ifndef PROJECT_NAME
    #define PROJECT_NAME            "Zv"
    #endif

    #define LED_PIN0 5
    #define MATRIX_WIDTH            600  // Number of pixels wide (how many LEDs per channel)
    #define MATRIX_HEIGHT           1   // Number of pixels tall
    #define NUM_LEDS                (MATRIX_WIDTH*MATRIX_HEIGHT)
    #define NUM_CHANNELS            1   // Everything wired sequentially on a single channel
    
    // #define NUM_FANS                10          // My system has 10 fans.  Because RGB. 1   // Our fans are on channels, not in sequential order, so only one "fan"
    // #define NUM_BANDS               8
    // #define NUM_RINGS               1           // Fans have a single outer ring of pixels // 5
    // #define RING_SIZE_0             24
    // #define BONUS_PIXELS            32          // Extra pixels - in this case, my case strip
    
    // #define COLOR_ORDER             EOrder::RGB
    #define ENABLE_AUDIO            0   // Listen for audio from the microphone and process it
    // #define POWER_LIMIT_MW       12 * 10 * 1000   // 10 amp supply at 5 volts assumed
    #define POWER_LIMIT_MW (INT_MAX)              // Unlimited power for long strips, up to you to limit here or supply enough!

    // Once you have a working project, selectively enable various additional features by setting
    // them to 1 in the list below.  This DEMO config assumes no audio (mic), or screen, etc.

    #ifndef ENABLE_WIFI
        #define ENABLE_WIFI             0   // Connect to WiFi
    #endif

    #define INCOMING_WIFI_ENABLED   1   // Accepting incoming color data and commands
    #define TIME_BEFORE_LOCAL       1   // How many seconds before the lamp times out and shows local content
    #define ENABLE_NTP              1   // Set the clock from the web
    #define ENABLE_OTA              1   // Accept over the air flash updates

    #if M5STICKC || M5STICKCPLUS
        #define LED_PIN0 32
    #elif LILYGOTDISPLAYS3
        #define LED_PIN0 21
    #else
        #define LED_PIN0 5
    #endif

    // The webserver serves files from its SPIFFS filesystem, such as index.html, and those files must be
    // uploaded to SPIFFS with the "Upload Filesystem Image" command before it can work.  When running
    // you should be able to see/select the list of effects by visiting the chip's IP in a browser.  You can
    // get the chip's IP by watching the serial output or checking your router for the DHCP given to 'LEDWifi'

    #ifndef ENABLE_WEBSERVER
        #define ENABLE_WEBSERVER        0   // Turn on the internal webserver
    #endif

    #define IR_REMOTE_PIN 15                        // Eric's is PIN 35 
    #define ENABLE_REMOTE           1   // IR Remote Control

    #define DEFAULT_EFFECT_INTERVAL     (MILLIS_PER_SECOND * 60 * 2)
    // #define DEFAULT_EFFECT_INTERVAL 1000 * 60 * 60 * 24     // One a day!

**Comment by davepl** [comment]
I’ve seen that call stack for a number of issues, it’s sort of the messenger, not the cause.

It looks like heap corruption (ie: a bug) but I don’t know the cause.  Oddly, I’ve found a hard reboot of the chip often solves it.  It’ll take 10 attempts to reboot but if you power cycle it, it comes right up.

Sorry I can’t be of more help!  Wish I knew...

**Comment by Metatron135** [comment]
Got it, will try, thanks Dave
